segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -120,295 +139,572 @@
-        // This step is kind of thread safe
+    /**
+     * Loads missing keys.
+     */
+    private CompletableFuture<Void> loadMissingKeys(MetadataTransaction txn, boolean skipStoreCheck, Map<String, TransactionData> txnData) {
+        val loadFutures = new ArrayList<CompletableFuture<TransactionData>>();
         for (Map.Entry<String, TransactionData> entry : txnData.entrySet()) {
-            String key = entry.getKey();
+            Preconditions.checkState(activeKeys.contains(entry.getKey()));
+            val key = entry.getKey();
             if (skipStoreCheck || entry.getValue().isPinned()) {
                 log.trace("Skipping loading key from the store key = {}", key);
             } else {
                 // This check is safe to be outside the lock
-                if (!bufferedTxnData.containsKey(key)) {
-                    loadFromStore(key);
+                val dataFromBuffer = bufferedTxnData.get(key);
+                if (null == dataFromBuffer) {
+                    loadFutures.add(loadFromStore(txn, key, true));
                 }
             }
         }
-        // Step 2 : Check whether transaction is safe to commit.
-        // This check needs to be atomic, with absolutely no possibility of re-entry
-        synchronized (lock) {
-            for (Map.Entry<String, TransactionData> entry : txnData.entrySet()) {
-                String key = entry.getKey();
-                val transactionData = entry.getValue();
-                Preconditions.checkState(null != transactionData.getKey());
-
-                // See if this entry was modified in this transaction.
-                if (transactionData.getVersion() == txn.getVersion()) {
-                    modifiedKeys.add(key);
-                    transactionData.setPersisted(false);
-                    modifiedValues.add(transactionData);
-                }
-                // make sure none of the keys used in this transaction have changed.
-                TransactionData dataFromBuffer = bufferedTxnData.get(key);
-                if (null != dataFromBuffer) {
-                    if (dataFromBuffer.getVersion() > transactionData.getVersion()) {
-                        throw new StorageMetadataVersionMismatchException(
-                                String.format("Transaction uses stale data. Key version changed key:%s buffer:%s transaction:%s",
-                                        key, dataFromBuffer.getVersion(), txnData.get(key).getVersion()));
+        return Futures.allOf(loadFutures)
+                .thenApplyAsync(v4 -> {
+                    // validate everything is alright.
+                    for (Map.Entry<String, TransactionData> entry : txnData.entrySet()) {
+                        val dataFromBuffer = bufferedTxnData.get(entry.getKey());
+                        if (!(entry.getValue().isPinned())) {
+                            Preconditions.checkState(activeKeys.contains(entry.getKey()));
```

Comment:
Can you please put a message with all these exceptions. If they fire and you see a stack in the logs you won't know where it came from. Put as much info in these messages as you can. It WILL help you in debugging these in the field.

Reply:
done.